arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -75,112 +72,95 @@ constexpr const char RETURN_TYPE_PARAM_NAME[] = "returnType";
-        arangodb::velocypack::deserializer::enum_member<
-            arangodb::iresearch::AnalyzerValueType::Object,
-            arangodb::velocypack::deserializer::values::string_value<
-                arangodb::iresearch::ANALYZER_VALUE_TYPE_OBJECT>>>;
-
-using OptionsDeserializer = utilities::constructing_deserializer<
-    Options,
-    parameter_list<
-        factory_deserialized_parameter<QUERY_STRING_PARAM_NAME,
-                                       values::value_deserializer<std::string>,
-                                       true>,
-        factory_simple_parameter<COLLAPSE_ARRAY_POSITIONS_PARAM_NAME, bool,
-                                 false, values::numeric_value<bool, false>>,
-        factory_simple_parameter<KEEP_NULL_PARAM_NAME, bool, false,
-                                 values::numeric_value<bool, true>>,
-        factory_simple_parameter<BATCH_SIZE_PARAM_NAME, uint32_t, false,
-                                 values::numeric_value<uint32_t, 10>>,
-        factory_simple_parameter<MEMORY_LIMIT_PARAM_NAME, uint32_t, false,
-                                 values::numeric_value<uint32_t, 1048576U>>,
-        factory_deserialized_default<
-            RETURN_TYPE_PARAM_NAME, AnalyzerValueTypeEnumDeserializer,
-            values::numeric_value<
-                arangodb::iresearch::AnalyzerValueType,
-                static_cast<std::underlying_type_t<
-                    arangodb::iresearch::AnalyzerValueType>>(
-                    arangodb::iresearch::AnalyzerValueType::String)>>>>;
-
-using ValidatingOptionsDeserializer =
-    validate<OptionsDeserializer, OptionsValidator>;
+template<class Inspector>
+inline auto inspect(Inspector& f,
+                    arangodb::iresearch::AqlAnalyzer::Options& o) {
+  using namespace arangodb::iresearch;
+  return f.object(o).fields(
+      f.field(QUERY_STRING_PARAM_NAME, o.queryString)
+          .invariant([](std::string const& v) -> arangodb::inspection::Status {
+            if (v.empty()) {
+              return {absl::StrCat("Value of '", QUERY_STRING_PARAM_NAME,
+                                   "' should be non empty string")};
+            }
+            return {};
+          }),
+      f.field(COLLAPSE_ARRAY_POSITIONS_PARAM_NAME, o.collapsePositions)
+          .fallback(false),
+      f.field(KEEP_NULL_PARAM_NAME, o.keepNull).fallback(true),
+      f.field(BATCH_SIZE_PARAM_NAME, o.batchSize)
+          .fallback(10u)
+          .invariant([](uint32_t v) -> arangodb::inspection::Status {
+            if (v == 0) {
+              return {absl::StrCat("Value of '", BATCH_SIZE_PARAM_NAME,
+                                   "' should be greater than 0")};
+            }
+            if (v > MAX_BATCH_SIZE) {
+              return {absl::StrCat("Value of '", BATCH_SIZE_PARAM_NAME,
+                                   "' should be less or equal to ",
+                                   MAX_BATCH_SIZE)};
+            }
+            return {};
+          }),
```

Comment:
I know it was, but will be nice to have another interface for deserialization

Something like 

```suggestion
      f.field(BATCH_SIZE_PARAM_NAME, o.batchSize)
          .fallback(10u)
          .booleanInvariant([](uint32_t v) -> bool {
            return v != 0 && v <= MAX_BATCH_SIZE;
          })
          .error([] () -> std::string {
            return absl::StrCat("Value of '", BATCH_SIZE_PARAM_NAME,
                                   "' should be greater than 0 and less or equal to ",
                                   MAX_BATCH_SIZE);
          },
```

And use this error not only in case of incorrect invariant, but also for not uint32_t values.

Reply:
Well, your last example not same. My main point is have custom message for case not uint32_t values.

Reply:
> the invariant from the actual error message, because you loose all the context.

For this place it's ok. here I don't need context to provide information to user what value should be specified

Reply:
If the value in the vpack is not a uint32 then you will automatically get a proper error message. That is one of the reasons why I implemented the inspector in the first place, so that deserializing from json produces better error messages. However, it will only contain the information which type was expected, which type we found, and which field (incl. path) we tried to deserialize. It is not possible to customize the error. Theoretically this could be added, but I would not want to do that in this PR.

Reply:
I don't think it's proper.
Proper write what you need to specify, this is impossible without custom error messqge

Reply:
I don't expect this will be changed in this PR, just some suggestions to the future.

Reply:
Changed to always initialize with 1.0